### PR TITLE
Controller: custom annotation for pinning ips (single and dual stack)

### LINKED ISSUE #1155: Custom annotation for pinning ips (single and dual stack)
The loadbalancerip field is gonna be deprecated (see https://github.com/kubernetes/kubernetes/pull/107235)

That has two implications:
- we need to have an alternative way to pin to a specific ip (the suggested way is to implement custom annotations on the service, see https://github.com/kubernetes/enhancements/pull/1992)
- that will enable us to pin both v6 and v4 addresses to the service (currently, only one is possible)

So, this issue is about having a custom annotation handled by metallb, that declares the user's will to use a specific ip (or ips).
See also https://github.com/kubernetes/enhancements/pull/1992#issuecomment-1000113208 for reference.

### PR DESCRIPTION
**Pull request by pperiyasamy**
This PR aims to fix #1155 issue by adding new custom annotation field `metallb.universe.tf/loadBalancerIPs `which can be used to request specific IPs (single and dual stack). For now this would coexist with `svc.Spec.LoadBalancerIP `field. when both fieds are specified, then annotation field takes higher precedence than spec field.

### CODE REVIEW

File: controller/controller_test.go [hunk 1]
```diff
@@ -543,6 +663,32 @@ func TestControllerMutation(t *testing.T) {
...
+						annotationLoadBalancerIPs: "1.2.3.0,1000::",
+					},
+				},
+				Spec: v1.ServiceSpec{
+					Type:       "LoadBalancer",
+					ClusterIPs: []string{"1.2.3.4", "3000::1"},
+				},
+				Status: statusAssigned([]string{"1.2.3.0", "1000::"}),
+			},
+		},
```
Comment (msherif1234) [inline controller/controller_test.go:691]: what if u have only one IP and we are in dual-stack mode would that flag an error ? can we add test coverage for this case too ?

File: controller/service.go [hunk 2]
```diff
@@ -214,3 +260,37 @@ func (c *controller) allocateIPs(key string, svc *v1.Service) ([]net.IP, error)
...
+		ip := net.ParseIP(desiredLbIP)
+		if ip == nil {
+			return nil, fmt.Errorf("invalid %s: %q", annotationLoadBalancerIPs, lbIPsStr)
+		}
+		lbIPs = append(lbIPs, ip)
+	}
+	return lbIPs, nil
+}
+
+func (c *controller) isLbIPsEqual(ipsA []net.IP, ipsB []net.IP) (bool, error) {
```
Comment (msherif1234) [inline controller/service.go:281]: why not using reflect.DeepEqual ?

File: controller/service.go [hunk 3]
```diff
@@ -121,6 +125,26 @@ func (c *controller) convergeBalancer(l log.Logger, key string, svc *v1.Service)
...
 		}
+		// annotationLoadBalancerIPs takes higher precedence over svc.Spec.LoadBalancerIP
+		desiredLbIPs, err := c.getDesiredLbIPs(svc)
+		if err != nil {
+			level.Error(l).Log("event", "loadbalancerIP", "error", err, "msg", "invalid requested loadbalancer IPs")
+			c.client.Errorf(svc, "LoadBalancerFailed", "invalid requested loadbalancer IPs: %s", err)
+			return true
+		}
+		if len(desiredLbIPs) > 0 {
+			ok, err := c.isLbIPsEqual(lbIPs, desiredLbIPs)
```
Comment (msherif1234) [inline controller/service.go:136]: can u use reflect.DeepEqual() instead of this new custom API ?

**Comment by msherif1234** [comment]
we probably need to add something here https://github.com/metallb/metallb/blob/main/website/content/usage/_index.md about this new annotation and that fact that loadbalancerIP will be deprecated

File: controller/service.go [hunk 4]
```diff
@@ -109,6 +111,8 @@ func (c *controller) convergeBalancer(l log.Logger, key string, svc *v1.Service)
 		// User set or changed the desired LB IP, nuke the
 		// state. allocateIP will pay attention to LoadBalancerIP and try
 		// to meet the user's demands.
+		// TODO: remove check for svc.Spec.LoadBalancerIP when this field is removed
+		// from service spec.
```
Comment (msherif1234) [inline controller/service.go:115]: since the annotation takes precedence over loadbalancerIp then there is no need to do both in this case and overwrite with annotation u can do the annotation 1st and skip loadbalancer check if len(desiredLbIps) > 0

File: controller/controller_test.go (hunk 1 above)
Comment (pperiyasamy) [reply controller/controller_test.go:706]: yes, it must throw an error as per the existing LoadBalancerIP unit test. added a test to cover that too.

File: controller/service.go (hunk 4 above)
Comment (pperiyasamy) [reply controller/service.go:116]: Good find ! changed it.

File: controller/service.go (hunk 3 above)
Comment (pperiyasamy) [reply controller/service.go:124]: at this point both lbIPs are unsorted, so couldn't use reflect.DeepEqual.

**Comment by pperiyasamy** [comment]
> we probably need to add something here https://github.com/metallb/metallb/blob/main/website/content/usage/_index.md about this new annotation and that fact that loadbalancerIP will be deprecated

done. please check.

File: controller/service.go [hunk 5]
```diff
@@ -106,10 +108,31 @@ func (c *controller) convergeBalancer(l log.Logger, key string, svc *v1.Service)
...
-		// state. allocateIP will pay attention to LoadBalancerIP and try
-		// to meet the user's demands.
-		if svc.Spec.LoadBalancerIP != "" {
+		// User set or changed the desired LB IP(s), nuke the
+		// state. allocateIP will pay attention to LoadBalancerIP(s) and try
+		// to meet the user's demands. annotationLoadBalancerIPs takes higher
+		// precedence over svc.Spec.LoadBalancerIP.
+		// TODO: remove check for svc.Spec.LoadBalancerIP when this field is removed
+		// from service spec.
+		desiredLbIPs, err := c.getDesiredLbIPs(svc)
```
Comment (fedepaol) [inline controller/service.go:117]: Can you have the LoadBalancerIP logic embedded in `getDesiredLbIPs` ? I think it will remove duplication and in general it will express what `getDesiredIPs` means.
So if the user set the annotation use that, otherwise fallback to Status.LoadBalancerIP .

File: controller/service.go [hunk 6]
```diff
@@ -106,10 +108,31 @@ func (c *controller) convergeBalancer(l log.Logger, key string, svc *v1.Service)
...
+		if err != nil {
```
Comment (fedepaol) [inline controller/service.go:118]: What's missing here is a check on ipFamily. If the user puts the wrong ip family compared to the service's one, we must fail.
I'd suggest incorporating that check to have `getDesiredLbIPs` return a slice and always check ipFamily. That will work also with the loadbalancerip type.

File: controller/service.go [hunk 7]
```diff
@@ -214,3 +257,37 @@ func (c *controller) allocateIPs(key string, svc *v1.Service) ([]net.IP, error)
...
 }
+
+func (c *controller) getDesiredLbIPs(svc *v1.Service) ([]net.IP, error) {
+	lbIPsStr := svc.Annotations[annotationLoadBalancerIPs]
+	if lbIPsStr == "" {
+		return nil, nil
+	}
+	var lbIPs []net.IP
+	desiredLbIPs := strings.Split(lbIPsStr, ",")
+	for _, desiredLbIP := range desiredLbIPs {
```
Comment (fedepaol) [inline controller/service.go:268]: We must ensure that if the ips are two, one is v4 and the other is v6.
Also, we must fail if there are more than 2.

File: controller/service.go [hunk 8]
```diff
@@ -214,3 +257,37 @@ func (c *controller) allocateIPs(key string, svc *v1.Service) ([]net.IP, error)
 	// Okay, in that case just bruteforce across all pools.
 	return c.ips.Allocate(key, serviceIPFamily, k8salloc.Ports(svc), k8salloc.SharingKey(svc), k8salloc.BackendKey(svc))
 }
+
+func (c *controller) getDesiredLbIPs(svc *v1.Service) ([]net.IP, error) {
```
Comment (fedepaol) [inline controller/service.go:261]: No need to be a method of controller.
This can be a pure function with no state.

File: controller/service.go [hunk 9]
```diff
@@ -214,3 +257,37 @@ func (c *controller) allocateIPs(key string, svc *v1.Service) ([]net.IP, error)
...
+}
+
+func (c *controller) isLbIPsEqual(ipsA []net.IP, ipsB []net.IP) (bool, error) {
+	if len(ipsA) > 2 || len(ipsB) > 2 {
+		return false, fmt.Errorf("unsupported number of load balancer IPs")
+	}
+	if len(ipsA) != len(ipsB) {
+		return false, nil
+	}
+	if len(ipsA) == 2 && (ipsA[0].String() == ipsB[0].String() || ipsA[0].String() == ipsB[1].String()) &&
```
Comment (fedepaol) [inline controller/service.go:285]: Can you just sort? It will be easier to understand.

File: controller/service.go [hunk 10]
```diff
@@ -214,3 +257,37 @@ func (c *controller) allocateIPs(key string, svc *v1.Service) ([]net.IP, error)
...
+		ip := net.ParseIP(desiredLbIP)
+		if ip == nil {
+			return nil, fmt.Errorf("invalid %s: %q", annotationLoadBalancerIPs, lbIPsStr)
+		}
+		lbIPs = append(lbIPs, ip)
+	}
+	return lbIPs, nil
+}
+
+func (c *controller) isLbIPsEqual(ipsA []net.IP, ipsB []net.IP) (bool, error) {
```
Comment (fedepaol) [inline controller/service.go:278]: Same, no need to be a method.

File: website/content/usage/_index.md [hunk 11]
```diff
@@ -20,6 +20,11 @@ address, or if the address is already in use by another service,
 assignment will fail and MetalLB will log a warning event visible in
 `kubectl describe service <service name>`.
 
+The `spec.loadBalancerIP` parameter is going to be deprecated soon, so
```
Comment (fedepaol) [inline website/content/usage/_index.md:23]: I'd reword this to:

"We both support `spec.loadBalancerIP` and a custom `metallb.universe.tf/loadBalancerIPs`.
`metallb.universe.tf/loadBalancerIPs` also supporst a comma separated list of IPs to be used in case of Dual Stack services.

Please note that `spec.LoadBalancerIP` is planned to be deprecated in [k8s apis](https://github.com/kubernetes/kubernetes/pull/107235).
"

File: controller/service.go [hunk 12]
```diff
@@ -214,3 +257,37 @@ func (c *controller) allocateIPs(key string, svc *v1.Service) ([]net.IP, error)
...
+	if len(ipsA) > 2 || len(ipsB) > 2 {
```
Comment (fedepaol) [inline controller/service.go:279]: This should be done in the `getDesiredLbIPs` function

File: controller/service.go (hunk 10 above)
Comment (fedepaol) [reply controller/service.go:278]: Also, rename as `ipsAreEqual` because that's what the function do.

File: controller/service.go (hunk 9 above)
Comment (fedepaol) [inline controller/service.go:285]: Also, you can use the `Equal` method of net.IP, no need to convert to string.

File: controller/service.go [hunk 13]
```diff
@@ -185,8 +208,24 @@ func (c *controller) allocateIPs(key string, svc *v1.Service) ([]net.IP, error)
 		return nil, err
 	}
 
-	// If the user asked for a specific IP, try that.
-	if svc.Spec.LoadBalancerIP != "" {
+	var lbIPs []net.IP
+	// annotationLoadBalancerIPs takes higher precedence over svc.Spec.LoadBalancerIP
+	// TODO: remove svc.Spec.LoadBalancerIP check when this field is removed
```
Comment (fedepaol) [inline controller/service.go:213]: Same, no need for TODO

File: controller/service.go [hunk 14]
```diff
@@ -185,8 +208,24 @@ func (c *controller) allocateIPs(key string, svc *v1.Service) ([]net.IP, error)
...
+	// from service spec.
+	desiredLbIPs, err := c.getDesiredLbIPs(svc)
```
Comment (fedepaol) [inline controller/service.go:215]: Same as above, incorporate `spec.LoadBalancerIP` in getDesiredLBIPs

File: controller/service.go [hunk 15]
```diff
@@ -185,8 +208,24 @@ func (c *controller) allocateIPs(key string, svc *v1.Service) ([]net.IP, error)
...
+	if err != nil {
+		return nil, err
+	}
+	if len(desiredLbIPs) > 0 {
+		lbipFamily, err := ipfamily.ForAddressesIPs(desiredLbIPs)
```
Comment (fedepaol) [inline controller/service.go:220]: nit: lbIPFamily. Also, rename to `desiredLbIPFamily`

**Comment by fedepaol** [comment]
Left a couple of comments, I'd also add an e2e test to make sure this works on the happy path.

File: controller/service.go [hunk 16]
```diff
@@ -106,10 +108,31 @@ func (c *controller) convergeBalancer(l log.Logger, key string, svc *v1.Service)
...
 		}
-		// User set or changed the desired LB IP, nuke the
-		// state. allocateIP will pay attention to LoadBalancerIP and try
-		// to meet the user's demands.
-		if svc.Spec.LoadBalancerIP != "" {
+		// User set or changed the desired LB IP(s), nuke the
+		// state. allocateIP will pay attention to LoadBalancerIP(s) and try
+		// to meet the user's demands. annotationLoadBalancerIPs takes higher
+		// precedence over svc.Spec.LoadBalancerIP.
+		// TODO: remove check for svc.Spec.LoadBalancerIP when this field is removed
```
[the remark this answers is not in the document]
Comment (pperiyasamy) [reply controller/service.go]: yes, done

File: controller/service.go (hunk 5 above)
Comment (pperiyasamy) [reply controller/service.go]: Yes! this is an elegant approach. made the changes now.

File: controller/service.go (hunk 6 above)
Comment (pperiyasamy) [reply controller/service.go:115]: done

File: controller/service.go (hunk 13 above)
Comment (pperiyasamy) [reply controller/service.go]: done

File: controller/service.go (hunk 14 above)
Comment (pperiyasamy) [reply controller/service.go]: done

File: controller/service.go (hunk 15 above)
Comment (pperiyasamy) [reply controller/service.go]: yes, done.

File: controller/service.go (hunk 8 above)
Comment (pperiyasamy) [reply controller/service.go]: of course, done

File: controller/service.go (hunk 7 above)
Comment (pperiyasamy) [reply controller/service.go]: The `ipfamily.ForAddressesIPs` method call already does the check, so we may not need an explicit check.

File: controller/service.go (hunk 10 above)
Comment (pperiyasamy) [reply controller/service.go]: removed this method and using `sort` and `reflect.DeepEqual `now

File: controller/service.go (hunk 12 above)
Comment (pperiyasamy) [reply controller/service.go]: this is no more applicable

File: controller/service.go (hunk 9 above)
Comment (pperiyasamy) [reply controller/service.go]: done

Comment (pperiyasamy) [reply controller/service.go]: ah! correct, anyway this is no more needed due to using `reflect.DeepEqual`

File: website/content/usage/_index.md (hunk 11 above)
Comment (pperiyasamy) [reply website/content/usage/_index.md]: done

File: e2etest/bgptests/bgp.go [hunk 17]
```diff
@@ -174,6 +174,11 @@ var _ = ginkgo.Describe("BGP", func() {
 				testservice.TrafficPolicyCluster(svc)
 				testservice.ForceV6(svc)
 			}),
+		table.Entry("IPV4 - ExternalTrafficPolicyCluster - request IPv4 via custom annotation", ipfamily.IPv4, "ExternalTrafficPolicyCluster", []string{v4PoolAddresses},
+			func(svc *corev1.Service) {
+				testservice.TrafficPolicyCluster(svc)
+				testservice.WithSpecificIP(svc, "192.168.10.100")
+			}),
```
Comment (msherif1234) [inline e2etest/bgptests/bgp.go:181]: pls add IPv6 and dual-stack coverage

**Comment by msherif1234** [comment]
you need to organize your commits I woudl suggest two commits one for the controller and controller-unittest and another for e2e coverage with proper title and descriptions

File: website/content/usage/_index.md [hunk 18]
```diff
@@ -20,6 +20,12 @@ address, or if the address is already in use by another service,
 assignment will fail and MetalLB will log a warning event visible in
 `kubectl describe service <service name>`.
 
+We both support `spec.loadBalancerIP` and a custom `metallb.universe.tf/loadBalancerIPs`
```
Comment (msherif1234) [inline website/content/usage/_index.md:23]: metalLB supports `spec.loadBalancerIP` and a custom `metallb.universe.tf/loadBalancerIPs`

File: e2etest/pkg/service/tweak.go [hunk 19]
```diff
@@ -25,3 +25,9 @@ func DualStack(svc *corev1.Service) {
 func TrafficPolicyCluster(svc *corev1.Service) {
 	svc.Spec.ExternalTrafficPolicy = corev1.ServiceExternalTrafficPolicyTypeCluster
 }
+
+func WithSpecificIP(svc *corev1.Service, ip string) {
```
Comment (msherif1234) [inline e2etest/pkg/service/tweak.go:29]: this can't be used to return multiple IPs to test dual-stack

File: e2etest/bgptests/bgp.go (hunk 17 above)
Comment (pperiyasamy) [reply e2etest/bgptests/bgp.go:187]: done

File: e2etest/pkg/service/tweak.go (hunk 19 above)
Comment (pperiyasamy) [reply e2etest/pkg/service/tweak.go]: yes, changed it to var-args now.

File: website/content/usage/_index.md (hunk 18 above)
Comment (pperiyasamy) [reply website/content/usage/_index.md]: done

**Comment by msherif1234** [comment]
pls take out the doc update from the 1st commit and have in its own commit

**Comment by pperiyasamy** [comment]
> pls take out the doc update from the 1st commit and have in its own commit

yes @msherif1234 , done.

File: e2etest/bgptests/bgp.go [hunk 20]
```diff
@@ -174,6 +174,17 @@ var _ = ginkgo.Describe("BGP", func() {
...
+		table.Entry("IPV4 - ExternalTrafficPolicyCluster - request IPv4 via custom annotation", ipfamily.IPv4, "ExternalTrafficPolicyCluster", []string{v4PoolAddresses},
+			func(svc *corev1.Service) {
+				testservice.TrafficPolicyCluster(svc)
+				testservice.WithSpecificIPs(svc, "192.168.10.100")
+			}),
+		table.Entry("DUALSTACK - ExternalTrafficPolicyCluster - request Dual Stack via custom annotation", ipfamily.DualStack, "ExternalTrafficPolicyCluster", []string{v4PoolAddresses, v6PoolAddresses},
+			func(svc *corev1.Service) {
+				testservice.TrafficPolicyCluster(svc)
+				testservice.DualStack(svc)
+				testservice.WithSpecificIPs(svc, "192.168.10.100", "fc00:f853:0ccd:e799::100")
```
Comment (fedepaol) [inline e2etest/bgptests/bgp.go:186]: Here we don't have control if the test will actually receive the ips we are asking for.
We only rely on the existing tests, which make sure the ips assigned to the service will work.

File: e2etest/pkg/service/tweak.go [hunk 21]
```diff
@@ -25,3 +29,19 @@ func DualStack(svc *corev1.Service) {
 func TrafficPolicyCluster(svc *corev1.Service) {
 	svc.Spec.ExternalTrafficPolicy = corev1.ServiceExternalTrafficPolicyTypeCluster
 }
+
+func WithSpecificIPs(svc *corev1.Service, ips ...string) {
+	if len(ips) == 0 {
+		return
+	}
+	var allIPs string
+	for _, ip := range ips {
```
Comment (fedepaol) [inline e2etest/pkg/service/tweak.go:38]: you can use `strings.Join`

File: e2etest/pkg/service/tweak.go [hunk 22]
```diff
@@ -25,3 +29,19 @@ func DualStack(svc *corev1.Service) {
...
+		allIPs += ip + ","
+	}
+	if len(allIPs) > 0 {
+		allIPs = strings.TrimRight(allIPs, ",")
+	}
+	svc.ObjectMeta.Annotations = map[string]string{
```
Comment (fedepaol) [inline e2etest/pkg/service/tweak.go:44]: ObjectMeta is embedded, you can use svc.Annotations

**Comment by pperiyasamy** [comment]
@fedepaol I've addressed most of your latest review comments, please have a look now.

File: e2etest/bgptests/bgp.go [hunk 23]
```diff
@@ -126,6 +126,19 @@ var _ = ginkgo.Describe("BGP", func() {
 
 			svc, _ := testservice.CreateWithBackend(cs, f.Namespace.Name, "external-local-lb", tweak)
 			defer testservice.Delete(cs, svc)
+			desiredLbIPs := svc.Annotations["metallb.universe.tf/loadBalancerIPs"]
+			if desiredLbIPs != "" {
+				ingressPts := svc.Status.LoadBalancer.Ingress
+				framework.ExpectNotEqual(len(ingressPts), 0)
```
Comment (fedepaol) [inline e2etest/bgptests/bgp.go:132]: no need for this check as you are comparing below.

File: e2etest/bgptests/bgp.go [hunk 24]
```diff
@@ -126,6 +126,19 @@ var _ = ginkgo.Describe("BGP", func() {
...
+				assignedIPs := strings.Join(func(ingresses []corev1.LoadBalancerIngress) []string {
```
Comment (fedepaol) [inline e2etest/bgptests/bgp.go:133]: This is a bit hard to follow. I'd just produce the []string out of svc.Status.LoadBalancer.Ingress and use it to feed strings.Join, or even better, do that in a `ValidateDesiredLB(svc)` func under the `service` package .

File: controller/service.go [hunk 25]
```diff
@@ -185,21 +186,20 @@ func (c *controller) allocateIPs(key string, svc *v1.Service) ([]net.IP, error)
...
-			return nil, fmt.Errorf("requested spec.loadBalancerIP %q does not match the ipFamily of the service", svc.Spec.LoadBalancerIP)
+	desiredLbIPs, desiredLbIPFamily, err := getDesiredLbIPs(svc)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the user asked for a specific IPs, try that.
+	if len(desiredLbIPs) > 0 {
+		if serviceIPFamily != desiredLbIPFamily {
+			return nil, fmt.Errorf("requested loadBalancer IP %q does not match the ipFamily of the service", desiredLbIPs)
```
Comment (fedepaol) [inline controller/service.go:197]: nit: loadBalancer IP(s)

Comment (pperiyasamy) [reply controller/service.go]: done

File: e2etest/bgptests/bgp.go (hunk 24 above)
Comment (pperiyasamy) [reply e2etest/bgptests/bgp.go]: sure, removed nested function now and also factored out this whole logic into `ValidateDesiredLB` function in` validate.go`

File: e2etest/bgptests/bgp.go (hunk 23 above)
Comment (pperiyasamy) [reply e2etest/bgptests/bgp.go]: yes, done.

**Comment by fedepaol** [comment]
Looks great, thanks!

**Comment by pperiyasamy** [comment]
Thanks @fedepaol and @msherif1234 for all the feedback !
